### PR TITLE
Issue 9 Solved

### LINKED ISSUE #9: 9. As a user, I want to filter my shopping list to make it easier to locate an item in the list.
AC:

* Display a text field above the top of the shopping list
* As the user types into the field, the list should narrow to display only items that contain the text the user entered in the filter field
* When the field has text in it, the user should be able to tap a UI element (e.g., with an "X" button next to the field) to clear the field
* The filter text should match any part of the item name (i.e. it should _not_ only match from the start of the string)

Notes:

* The wireframe shows the filter field, but not it being used. https://github.com/the-collab-lab/smart-shopping-list/blob/main/_resources/wireframes/list-populated.png
* The following example app from MDN is a great example of providing feedback that is accessible to screenreaders when things change on the screen. https://moz-todomvc-react.netlify.app/

### PR DESCRIPTION
**Pull request by EmilyDSarani**
## Description

We added a  .filter that will allow users to search through their list using a text input. If the user types in 'cookie', then anything with  the word 'cookie' comes up.  We normalized the input field to lower case, so the user could potentially type 'Cookie' or 'COOKIE' and the same words with cookie in it will still populate in the list.  We also created a button by the text input that allows the user to clear out their inputs at anytime. 

## Related Issue

Closes #9 

## Type of Changes

- [x] Display a text field above the top of the shopping list
- [x] As the user types into the field, the list should narrow to display only items that contain the text the user entered in the filter field
- [x] When the field has text in it, the user should be able to tap a UI element (e.g., with an "X" button next to the field) to clear the field
- [x] The filter text should match any part of the item name (i.e. it should not only match from the start of the string)

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

ScreenShot of List Filter from Tim

## Testing Steps / QA Criteria

```
git pull origin sb-es-tt-list-filter
npm start
```

### CODE REVIEW

File: src/components/AddItem/AddItem.js [hunk 1]
```diff
@@ -126,3 +126,6 @@ const AddItem = () => {
 };
 
 export default AddItem;
+
```
Comment (An-dy1) [inline src/components/AddItem/AddItem.js]: can this be deleted, or are you saving it for the future?

File: src/components/List/List.js [hunk 2]
```diff
@@ -18,29 +19,46 @@ export const timeDifference = (timestampInMilli) => {
   const difference = date - timestampInMilli;
   const comparison = difference < dayInMilli;
```
Comment (An-dy1) [inline src/components/List/List.js:20]: Did you elect to comment out to keep the console clear, but retain these in comments for future reference? I'd recommend deleting at this point but if they're useful, tell me so and feel free to keep them around.

File: src/components/List/List.js [hunk 3]
```diff
@@ -18,29 +19,46 @@ export const timeDifference = (timestampInMilli) => {
...
-    // 2: purchaseDate exists, checkbox is checked => unchecking sets purchaseDate to null
-    // 3: purchaseDate exists, checkbox is unchecked => checking runs serverTimestamp
+    // 1) purchaseDate is null, checkbox is unchecked => checking runs servertimestamp
+    // 2) purchaseDate exists, checkbox is checked => unchecking sets purchaseDate to null
+    // 3) purchaseDate exists, checkbox is unchecked => checking runs serverTimestamp
 
     await updateDoc(doc(db, token, id), {
       purchaseDate: checked ? null : serverTimestamp(),
     });
   }
```
Comment (An-dy1) [inline src/components/List/List.js]: hmm - can this be deleted here? This must be code for issue 10, right? I'm anticipating that keeping these lines commented out here could create a merge conflict if this one is merged in as-is, with the un-commented code in the other PR.

File: src/components/List/List.js [hunk 4]
```diff
@@ -74,21 +92,49 @@ const List = () => {
...
+                // item will pass through filter
+                //
+                // 2) if search string is part of item name
+                // item will pass through filter
+                //
+                // 3) otherwise item will be filtered out
+                if (userSearch === '') {
+                  return true;
+                } else if (
+                  doc.item.toLowerCase().includes(userSearch.toLowerCase())
```
Comment (An-dy1) [inline src/components/List/List.js:93]: nice - way to take care of case-insensitivity

**Review by An-dy1 — CHANGES_REQUESTED** [review]

**Comment by timothy-taylor** [comment]
@An-dy1, all of the comments and console log things have been cleaned up in the Issue 10 branch. I'm happy to change them here too, but it should get taken care of in the merge. Let me know what you think.

edit: I realized I didn't have the latest version on my machine. Didn't notice that some code from issue 10 ended up in here. Fixed all above issues.

**Comment by An-dy1** [comment]
Thanks Tim - I hadn't looked at issue 10 yet, so that makes sense. Approved!

**Review by An-dy1 — APPROVED** [review]

**Pull request merged**